cpp/src/arrow/compute/kernels/codegen_internal.cc

```diff
@@ -109,6 +109,39 @@ void ReplaceNullWithOtherType(ValueDescr* first, size_t count) {
   }
 }
 
+void ReplaceTemporalTypes(const TimeUnit::type unit, std::vector<ValueDescr>* descrs) {
+  auto* end = descrs->data() + descrs->size();
+  const std::string* timezone = nullptr;
+
+  for (auto* it = descrs->data(); it != end; it++) {
+    switch (it->type->id()) {
+      case Type::TIMESTAMP: {
+        const auto& ty = checked_cast<const TimestampType&>(*it->type);
+        timezone = &ty.timezone();
+        if (timezone) {
+          it->type = timestamp(unit, *timezone);
+        } else {
+          it->type = timestamp(unit);
+        }
+        continue;
+      }
+      case Type::TIME32:
+      case Type::TIME64:
+      case Type::DURATION: {
+        it->type = duration(unit);
+        continue;
+      }
+      case Type::DATE32:
+      case Type::DATE64: {
+        it->type = timestamp(unit);
```

Comment:
Wait, shouldn't this be `date32`/`date64`?

Reply:
Ah, okay. I guess this is kind of specific to temporal arithmetic.

Reply:
Yeah this is for arithmetic purposes.